src/render/quantum.rs

```diff
@@ -1,75 +1,46 @@
 //! Optimized audio signal data structures, used in `AudioProcessors`
 use arrayvec::ArrayVec;
-use std::cell::RefCell;
 use std::rc::Rc;
 
 use crate::node::{ChannelConfigInner, ChannelCountMode, ChannelInterpretation};
 
 use crate::assert_valid_number_of_channels;
 use crate::{MAX_CHANNELS, RENDER_QUANTUM_SIZE};
 
-// object pool for `AudioRenderQuantumChannel`s, only allocate if the pool is empty
-pub(crate) struct Alloc {
-    inner: Rc<AllocInner>,
-}
-
-#[derive(Debug)]
-struct AllocInner {
-    pool: RefCell<Vec<Rc<[f32; RENDER_QUANTUM_SIZE]>>>,
-    zeroes: Rc<[f32; RENDER_QUANTUM_SIZE]>,
-}
-
-impl Alloc {
-    pub fn with_capacity(n: usize) -> Self {
-        let pool: Vec<_> = (0..n).map(|_| Rc::new([0.; RENDER_QUANTUM_SIZE])).collect();
-        let zeroes = Rc::new([0.; RENDER_QUANTUM_SIZE]);
+/// A sample slice containing only zeroes
+const ZEROES: [f32; RENDER_QUANTUM_SIZE] = [0.; RENDER_QUANTUM_SIZE];
 
-        let inner = AllocInner {
-            pool: RefCell::new(pool),
-            zeroes,
-        };
-
-        Self {
-            inner: Rc::new(inner),
-        }
-    }
+mod pool {
+    //! Thread local allocator for sample slices
 
-    #[cfg(test)]
-    pub fn allocate(&self) -> AudioRenderQuantumChannel {
-        AudioRenderQuantumChannel {
-            data: self.inner.allocate(),
-            alloc: Rc::clone(&self.inner),
-        }
-    }
+    use super::*;
+    use std::cell::RefCell;
 
-    pub fn silence(&self) -> AudioRenderQuantumChannel {
-        AudioRenderQuantumChannel {
-            data: Rc::clone(&self.inner.zeroes),
-            alloc: Rc::clone(&self.inner),
-        }
+    thread_local! {
+        /// Thread local pool of pre-allocated sample slice that can be reused after being dropped
+        static POOL: RefCell<Vec<Rc<[f32; RENDER_QUANTUM_SIZE]>>> = RefCell::new(Vec::with_capacity(32));
```

Comment:
I wonder if we shouldn't have a larger initial pool? As a stereo delay of 1 second (default `max_delay_time` uses 375 * 2 `AudioRenderQuantumChannel`, having something like 1024 preallocated buffers would guarantees we don't allocate at least when creating one delay

Reply:
Good point. We could definitely do it but we should realize this could happen on any thread and the memory is never reclaimed, so to be a bit conservative would not be bad.
I'm having a bit more thoughts about this setup though, when using async rendering the OfflineAudioContext can move to any thread of the executor and we may end up leaking a lot of memory.
I could specialize it to be of different size for offline and online rendering. Starting with a small pool for offline rendering is not bad because we can safely allocate more during rendering.